artifact/version/version.cpp

```diff
@@ -0,0 +1,124 @@
+
+#include <string>
+#include <vector>
+
+#include <common/common.hpp>
+#include <common/expected.hpp>
+#include <common/error.hpp>
+#include <common/json.hpp>
+
+#include <artifact/version/version.hpp>
+
+namespace artifact {
+namespace v3 {
+namespace version {
+
+namespace io = mender::common::io;
+namespace expected = mender::common::expected;
+namespace error = mender::common::error;
+namespace json = mender::common::json;
+
+
+const ErrorCategoryClass ErrorCategory {};
+
+const char *ErrorCategoryClass::name() const noexcept {
+	return "VersionParserErrorCategory";
+}
+
+string ErrorCategoryClass::message(int code) const {
+	switch (code) {
+	case NoError:
+		return "Success";
+	ParseError:
+		return "Parse error";
+	VersionError:
+		return "Wrong Artifact version";
+	FormatError:
+		return "Wrong Artifact format";
+	default:
+		return "Unknown";
+	}
+}
+
+error::Error MakeError(ErrorCode code, const string &msg) {
+	return error::Error(error_condition(code, ErrorCategory), msg);
+}
+
+
+expected::Expected<Version, error::Error> Parse(io::Reader &reader) {
+	// Read in all the bytes into a tmp buffer
+	std::vector<uint8_t> buf(1024);
+	io::ByteWriter bw {buf};
+
+	auto err = io::Copy(bw, reader);
+	if (err) {
+		return err;
+	}
+
+	auto expected_json = json::LoadFromString(mender::common::StringFromByteVector(buf));
+
+	if (!expected_json) {
+		return MakeError(ParseError, expected_json.error().message);
+	}
+
+	const json::ExpectedJson expected_version = expected_json.value();
+
+	if (!expected_version) {
+		return MakeError(ParseError, expected_json.error().message);
+	}
+
+	auto expected_obj = expected_version.value();
```

Comment:
I'm somewhat confused by this code. Is there three layers of ExpectedJson inside each other? Line 64 looks like a no-op, you are just assigning a payload value to an ExpectedJson, and since it is not an error, it will always construct an ExpectedJson with a value, and the next `if` will always evaluate false.

But feel free to enlighten me if I'm misunderstanding this.

Reply:
I don't even know. I think this is horrible...

However, I have not come up with a cleaner way...

Reply:
Alright, cleaning it up a little...